MatrixSDK/MXSession.m

```diff
@@ -1512,7 +1512,18 @@ - (void)handleServerSyncError:(NSError*)error forRequestWithServerTimeout:(NSUIn
         if ([error.domain isEqualToString:NSURLErrorDomain]
             && code == kCFURLErrorCancelled)
         {
-            MXLogDebug(@"[MXSession] The connection has been cancelled.");
+            MXLogDebug(@"[MXSession] The connection has been cancelled in state %@", @(_state));
+
+            if (_state == MXSessionStateSyncInProgress)
+            {
+                // This happens when the SDK cannot make any more requests because the app is in background
+                // and the background task is expired or going to expire.
+                // The app should have paused the SDK before but it did not. So, pause the SDK ourselves.
+                // Note that we need to come back to MXSessionStateRunning in order to be able to pause.
+                MXLogDebug(@"[MXSession] -> Go to pause");
+                [self setState:MXSessionStateRunning];
```

Comment:
Instead of going into _running_ state, `MXSessionStatePauseRequested` makes more sense to me, in terms of state machine logically, which should also be applicable to pause the session.

I'm just imagining that if we cannot pause the session for some reason, we'll stay in running state and next time there will be a similar issue when we tried to resume the session, we'll see the session is already _running_ and probably do nothing.